### PR TITLE
UCP/PROTO: Add generic completion callback for ucp_proto request.

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

File: src/ucp/proto/proto_am.c [hunk 1]
```diff
@@ -42,7 +42,7 @@ ucs_status_t ucp_proto_progress_am_bcopy_single(uct_pending_req_t *self)
     ucs_status_t status = ucp_do_am_bcopy_single(self, req->send.proto.am_id,
                                                  ucp_proto_pack);
     if (status == UCS_OK) {
-        ucp_request_put(req);
+        req->send.proto.comp_cb(req);
```
Comment (hoopoepg) [inline src/ucp/proto/proto_am.c]: I believe we never come here from non-am branch, else we have troubles calling uninitialized pointer to function

Comment (yosefe) [reply src/ucp/proto/proto_am.c]: this is a preparation for allowing other protocol (e.g rndv) setting custom completion callbacks

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5483/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5487/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5489/ for details (Mellanox internal link).

**Pull request merged**